modules/dex/dex.go

```diff
@@ -0,0 +1,347 @@
+package dex
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	_ "embed"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	dexapi "github.com/dexidp/dex/api/v2"
+	"github.com/google/uuid"
+	"golang.org/x/crypto/bcrypt"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+const (
+	HTTPPort = "5556/tcp"
+	GRPCPort = "5557/tcp"
+)
+
+const (
+	bcryptCost    = 10
+	defaultIssuer = "http://localhost:5556"
+)
+
+var ErrPasswordRequired = errors.New("plaintext password or hash is required")
+
+var (
+	propertiesToPatch = []string{
+		"authorization_endpoint",
+		"token_endpoint",
+		"jwks_uri",
+		"userinfo_endpoint",
+		"device_authorization_endpoint",
+		"introspection_endpoint",
+	}
+	//go:embed config/dex-config.yaml.tmpl
+	dexConfigTemplate []byte
+)
+
+// CreateClientAppRequest represents a request to register a client application in Dex.
+type CreateClientAppRequest struct {
+	Name         string
+	ClientID     string
+	ClientSecret string
+	RedirectURIs []string
+	Public       bool
+}
+
+// CreatePasswordRequest represents a request to register an identity in Dex.
+// Either Hash or Password must be provided.
+// If Hash is provided, Password will be ignored.
+type CreatePasswordRequest struct {
+	UserID   string
+	Email    string
+	Username string
+	// Password is a plain text password.
+	Password string
+	// Hash is a bcrypt hash of the password.
+	Hash []byte
+}
+
+// Container represents the Dex container type used in the module
+type Container struct {
+	testcontainers.Container
+	Client *http.Client
+}
+
+// CreateClientApp registers a new client application in Dex.
+// If ClientID or ClientSecret are empty, they will be generated and returned.
+func (c Container) CreateClientApp(ctx context.Context, req CreateClientAppRequest) (*dexapi.Client, error) {
+	apiClient, connCloser, err := c.createDexAPIClient(ctx)
+	if err != nil {
+		return nil, err
```

Comment:
suggestion: wrap the error so we know where it came from, more below and in other files.

Reply:
Thanks for the reminder! 🙈
Apparently, I rushed a bit too fast through the prototype 😅